File: wallet/ios/Core/Sources/Core/HubAuthService.swift
import Foundation
import TweetNacl

/// Sign-In-With-Solana against a Kvasir hub to obtain the bearer token an
/// autonomous node uses to poll/enroll on an auth-gated (public, remote) hub.
///
/// Mirrors `wallet/android/.../HubAuthService.kt`. Flow (controller/siws.py,
/// controller/hub.py):
///   1. POST /api/auth/challenge {wallet}          -> {nonce, message}
///   2. sign the message bytes with the wallet key -> base64 ed25519 signature
///   3. POST /api/auth/node-token {wallet,nonce,signature} -> {node_token}
/// The node token is scoped to participation, so the hub skips 2FA (which a
/// mobile wallet has no UI for). Returns the bearer token the node polls with.
public actor HubAuthService {
    private let baseString: String
    private let mnemonic: [String]
    private let session = URLSession(configuration: .ephemeral)

    public init(baseUrl: String, mnemonic: [String]) {
        var b = baseUrl.trimmingCharacters(in: .whitespacesAndNewlines)
        while b.hasSuffix("/") { b.removeLast() }
        self.baseString = b
        self.mnemonic = mnemonic.map { $0.lowercased() }
    }

    private struct Challenge: Decodable { let nonce: String?; let message: String? }
    private struct TokenResp: Decodable { let node_token: String? }

    /// Mint a long-lived NODE token with a wallet signature alone — no OTP.
    public func nodeToken() async throws -> String {
        let kp = try await WalletDeriver.keyPair(phrase: mnemonic)
        let wallet = kp.publicKey.base58EncodedString

        let ch: Challenge = try await post("/api/auth/challenge", ["wallet": wallet])
        guard let message = ch.message, let nonce = ch.nonce,
              !message.isEmpty, !nonce.isEmpty else {
            throw HubAuthError.message("hub did not issue a challenge")
        }
        let sig = try NaclSign.signDetached(message: Data(message.utf8), secretKey: kp.secretKey)
        let sigB64 = sig.base64EncodedString()

        let resp: TokenResp = try await post("/api/auth/node-token",
            ["wallet": wallet, "nonce": nonce, "signature": sigB64])
        guard let token = resp.node_token, !token.isEmpty else {
            throw HubAuthError.message("hub issued no node token")
        }
        return token
    }

    // MARK: transport (mirrors StakingService)
    private func post<T: Decodable>(_ path: String, _ body: [String: Any]) async throws -> T {
        guard let url = URL(string: baseString + path) else { throw HubAuthError.message("bad hub URL") }
        var req = URLRequest(url: url)
        req.httpMethod = "POST"
        req.timeoutInterval = 20
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.httpBody = try JSONSerialization.data(withJSONObject: body)
        let (data, resp) = try await session.data(for: req)
        let code = (resp as? HTTPURLResponse)?.statusCode ?? 0
        guard (200..<300).contains(code) else {
            let msg = (try? JSONDecoder().decode([String: String].self, from: data))?["error"]
                ?? String(data: data, encoding: .utf8) ?? "HTTP \(code)"
            throw HubAuthError.message(msg)
        }
        return try JSONDecoder().decode(T.self, from: data)
    }
}

public enum HubAuthError: Error, CustomStringConvertible {
    case message(String)
    public var description: String { if case let .message(m) = self { return m }; return "hub auth error" }
}

File: wallet/android/app/src/main/kotlin/ai/banya/linkcpp/wallet/HubAuthService.kt
package ai.banya.linkcpp.wallet

import ai.banya.linkcpp.core.WalletDeriver
import android.util.Base64
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL

/**
 * Sign-In With Solana against a Kvasir hub, to obtain the bearer token an
 * autonomous node uses to poll/enroll on an auth-gated (public, remote) hub.
 *
 * Flow (see controller/siws.py, controller/hub.py):
 *   1. POST /api/auth/challenge {wallet}         -> {nonce, message}
 *   2. sign message bytes with the wallet key    -> base64 signature (ed25519)
 *   3. POST /api/auth/verify {wallet,nonce,sig}  -> session cookie, OR
 *                                                   {twofa_required, pre_auth}
 *   4. POST /api/auth/2fa/login {pre_auth,code}  -> session cookie
 * The session cookie value (linkcpp_session) doubles as the Authorization
 * bearer token the hub accepts (_authed_wallet).
 */
class HubAuthService(baseUrl: String, private val mnemonic: List<String>) {
    private val base = baseUrl.trimEnd('/')
    private val sessionCookie = "linkcpp_session"

    private val keypair get() = WalletDeriver.keypair(mnemonic)
    val wallet: String get() = keypair.publicKey.toBase58()

    /**
     * Mint a long-lived NODE token with a wallet signature alone — no OTP.
     * The node token is scoped to participation, so the hub skips 2FA (which a
     * mobile wallet has no UI for). Returns the bearer token the node polls with.
     */
    fun nodeToken(): String {
        val kp = keypair
        val w = kp.publicKey.toBase58()
        val ch = postJson("/api/auth/challenge", JSONObject().put("wallet", w))
        val message = ch.optString("message"); val nonce = ch.optString("nonce")
        if (message.isEmpty() || nonce.isEmpty()) error("hub did not issue a challenge")
        val sigB64 = Base64.encodeToString(kp.sign(message.toByteArray(Charsets.UTF_8)), Base64.NO_WRAP)
        val resp = postJson("/api/auth/node-token",
            JSONObject().put("wallet", w).put("nonce", nonce).put("signature", sigB64))
        return resp.optString("node_token").ifEmpty { error("hub issued no node token") }
    }

    private fun postJson(path: String, body: JSONObject): JSONObject {
        val conn = (URL(base + path).openConnection() as HttpURLConnection).apply {
            requestMethod = "POST"; connectTimeout = 12000; readTimeout = 20000; doOutput = true
            setRequestProperty("Content-Type", "application/json")
        }
        try {
            conn.outputStream.use { it.write(body.toString().toByteArray(Charsets.UTF_8)) }
            val code = conn.responseCode
            val text = (if (code in 200..299) conn.inputStream else conn.errorStream)
                ?.bufferedReader()?.use { it.readText() } ?: ""
            val obj = runCatching { JSONObject(text) }.getOrDefault(JSONObject())
            if (code !in 200..299) error(obj.optString("error", "HTTP $code"))
            return obj
        } finally {
            conn.disconnect()
        }
    }
}
